Dockerfile

```diff
@@ -1,7 +1,30 @@
-FROM alpine as alpine
-RUN apk add -U --no-cache ca-certificates
+FROM golang:1.18.3-alpine3.16@sha256:7cc62574fcf9c5fb87ad42a9789d5539a6a085971d58ee75dd2ee146cb8a8695 AS build
 
-FROM scratch
-COPY --from=alpine /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
-ENTRYPOINT ["/dnscontrol"]
-COPY dnscontrol /
+WORKDIR /go/src/github.com/StackExchange/dnscontrol
+
+ARG BUILD_VERSION
+
+ENV GO111MODULE on
+
+COPY . .
+
+# build dnscontrol
+RUN apk update \
+    && apk add --no-cache ca-certificates curl gcc build-base git \
+    && update-ca-certificates \
+    && go build -v -trimpath -buildmode=pie -ldflags="-s -w -X main.SHA=${BUILD_VERSION}"
+
+# Validation check
+RUN cp dnscontrol /go/bin/dnscontrol
+RUN dnscontrol version
+
+# -----
+
+FROM alpine:3.16.0@sha256:686d8c9dfa6f3ccfc8230bc3178d23f84eeaf7e457f36f271ab1acc53015037c
```

Comment:
@fdcastel why not from scratch?

Reply:
I don't know. To the best of my understanding `scratch` is an empty image more appropriate for base O.S. images (like alpine, debian, etc.)

I have no experience with `FROM scratch` images. Opinions?

Reply:
the dnscontrol go binary shouldn't have any external dependencies and therefore doesn't need a base os in the docker container, therefore it is sufficient to use the scratch image.

Reply:
I see.  @jauderho Opinions?

P.S.: A brief search pointed me this:

https://ashishb.net/all/docker-be-careful-about-the-scratch-image/

Dnscontrol doesn't have any static dependencies, but we may need some external services (e.g. Cloudflare, via https) in runtime.

I cannot test it right now but will asap.

Reply:
@fdcastel yes the COPY line should be there in any case
`COPY --from=alpine /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/`

Reply:
@aba-rechsteiner Have you actually tried to build this with scratch as the base image? It doesn't work for me which is why I settled on using alpine.